### ISSUE TITLE
FP in if_same_then_else for if-let

### ISSUE THREAD

**Issue by extrawurst** [rust-lang/rust-clippy#7579]
followup to #7383 and pinging @xFrednet

Lint name: `if_same_then_else`

I tried this code:

```rust
#[allow(unused_variables)]
fn main() {
    let a = true;
    let b = Some(true);
    
    if let Some(b) = b {
        println!("hello");
    }
    else if a {
        println!("hello");
    }
}
```
see https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=c49e249051ac25681b4fe4522994a524

I expected to see this happen: 
no lint warning, since I cannot express this until *if-let-chains* are supported

Instead, this happened: 
```sh
Checking playground v0.0.1 (/playground)
error: this `if` has identical blocks
  --> src/main.rs:6:24
   |
6  |       if let Some(b) = b {
   |  ________________________^
7  | |         println!("hello");
8  | |     }
   | |_____^
   |
   = note: `#[deny(clippy::if_same_then_else)]` on by default
note: same as this
  --> src/main.rs:9:15
   |
9  |       else if a {
   |  _______________^
10 | |         println!("hello");
11 | |     }
   | |_____^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#if_same_then_else

error: could not compile `playground` due to previous error
```

### Meta

- `cargo clippy -V`: clippy 0.1.56 (0035d9dc 2021-08-16)
- `rustc -Vv`:
```
rustc 1.54.0 (a178d0322 2021-07-26)
binary: rustc
commit-hash: a178d0322ce20e33eac124758e837cbd80a6f633
commit-date: 2021-07-26
host: x86_64-apple-darwin
release: 1.54.0
LLVM version: 12.0.1
```

**Comment by camsteffen** [comment]
This is a regression from https://github.com/rust-lang/rust/pull/80357.

**Comment by flip1995** [comment]
There were [allows added](https://github.com/rust-lang/rust/pull/80357/files?file-filters%5B%5D=.diff&file-filters%5B%5D=.mir&file-filters%5B%5D=.rs&file-filters%5B%5D=.stderr#diff-fc2911f28b15d623533677d5d2f4947ee9692ec257dafb20fdb00dd1f8fa03fbR7) for this lint in our tests in rust-lang/rust#80357

These may have been added to get around this bug :thinking:
